pkg/tsdb/cloudwatch/time_series_query_test.go

```diff
@@ -136,3 +137,268 @@ func TestTimeSeriesQuery(t *testing.T) {
+	im := datasource.NewInstanceManager(func(s backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		return datasourceInfo{}, nil
+	})
+	executor := newExecutor(im, newTestConfig(), &fakeSessionCache{})
+
+	resp, err := executor.QueryData(context.Background(), &backend.QueryDataRequest{
+		PluginContext: backend.PluginContext{DataSourceInstanceSettings: &backend.DataSourceInstanceSettings{}},
+		Queries: []backend.DataQuery{
+			{
+				RefID:     "A",
+				TimeRange: backend.TimeRange{From: time.Now().Add(time.Hour * -2), To: time.Now().Add(time.Hour * -1)},
+				JSON: json.RawMessage(`{
+						"type":      "timeSeriesQuery",
+						"metricQueryType": 0,
+						"metricEditorMode": 1,
+						"namespace": "",
+						"metricName": "",
+						"expression": "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 300)",
+						"region": "us-east-2",
+						"id": "a",
+						"alias": "{{period}} {{stat}}",
+						"statistic": "Maximum",
+						"period": "1200",
+						"hide": false,
+						"matchExact": true,
+						"refId": "A"
+					}`),
+			},
+		},
+	})
+
+	assert.NoError(t, err)
+	// asserts that period '300' and stat 'Average' are parsed from the JSON query expression and used in the alias, not the JSON model's 'statistic' nor 'period' fields
+	assert.Equal(t, "300 Average", resp.Responses["A"].Frames[0].Name)
+}
+
+func Test_QueryData_executeTimeSeriesQuery_no_alias_provided_frame_name_is_queryId_when_query_isMathExpression(t *testing.T) {
+	origNewCWClient := NewCWClient
+	t.Cleanup(func() {
+		NewCWClient = origNewCWClient
+	})
+	var cwClient fakeCWClient
+	NewCWClient = func(sess *session.Session) cloudwatchiface.CloudWatchAPI {
+		return &cwClient
+	}
+	cwClient = fakeCWClient{
+		GetMetricDataOutput: cloudwatch.GetMetricDataOutput{
+			MetricDataResults: []*cloudwatch.MetricDataResult{
+				{StatusCode: aws.String("Complete"), Id: aws.String("query id"), Label: aws.String("NetworkOut"),
+					Values: []*float64{aws.Float64(1.0)}, Timestamps: []*time.Time{{}}},
+			},
+		},
+	}
+	im := datasource.NewInstanceManager(func(s backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+		return datasourceInfo{}, nil
+	})
+	executor := newExecutor(im, newTestConfig(), &fakeSessionCache{})
+
+	resp, err := executor.QueryData(context.Background(), &backend.QueryDataRequest{
+		PluginContext: backend.PluginContext{DataSourceInstanceSettings: &backend.DataSourceInstanceSettings{}},
+		Queries: []backend.DataQuery{
+			{
+				RefID:     "A",
+				TimeRange: backend.TimeRange{From: time.Now().Add(time.Hour * -2), To: time.Now().Add(time.Hour * -1)},
+				JSON: json.RawMessage(`{
+						"type":      "timeSeriesQuery",
+						"metricQueryType": 0,
+						"metricEditorMode": 1,
+						"namespace": "",
+						"metricName": "",
+						"region": "us-east-2",
+						"id": "query id",
+						"statistic": "Maximum",
+						"period": "1200",
+						"hide": false,
+						"matchExact": true,
+						"refId": "A"
+					}`),
```

Comment:
Hmm if it’s a huge vertical block JSON objects in a separate file, might be hard to compare there too...

When adding a batch of tests like this without changes in the code that enter those scenarios, I think it might be challenging to digest every test.
Imagine if each test here had been added alongside the PR which added the functionality. Then I think that the tests would have been more approachable to review and we wouldn’t be trying to compare different scenarios side by side, but instead we’d be focused on one scenario and trying to figure out if enough tests are written around that one scenario. 

I’ll try to think of another way to improve this.